File: app/domain/normalization.py
from __future__ import annotations

import re
from enum import StrEnum

CYRILLIC_TO_LATIN = str.maketrans(
    {
        "А": "A",
        "В": "B",
        "Е": "E",
        "І": "I",
        "К": "K",
        "М": "M",
        "Н": "H",
        "О": "O",
        "Р": "P",
        "С": "C",
        "Т": "T",
        "Х": "X",
    }
)
PLATE_RE = re.compile(r"^[ABCEHIKMOPTX]{2}\d{4}[ABCEHIKMOPTX]{2}$")
VIN_RE = re.compile(r"^[A-HJ-NPR-Z0-9]{17}$")


class QueryKind(StrEnum):
    PLATE = "PLATE"
    VIN = "VIN"
    UNKNOWN = "UNKNOWN"


def normalize_plate(value: str) -> str | None:
    candidate = re.sub(r"[\s\-]+", "", value.strip().upper()).translate(CYRILLIC_TO_LATIN)
    return candidate if PLATE_RE.fullmatch(candidate) else None


def normalize_vin(value: str) -> str | None:
    candidate = re.sub(r"[\s\-]+", "", value.strip().upper())
    return candidate if VIN_RE.fullmatch(candidate) else None


def detect_query(value: str) -> tuple[QueryKind, str | None]:
    vin = normalize_vin(value)
    if vin:
        return QueryKind.VIN, vin
    plate = normalize_plate(value)
    if plate:
        return QueryKind.PLATE, plate
    return QueryKind.UNKNOWN, None

File: worker/src/normalization.ts
import type { Language } from "./types.js";

const CYRILLIC_TO_LATIN: Record<string, string> = {
  "А": "A", "В": "B", "Е": "E", "І": "I", "К": "K", "М": "M",
  "Н": "H", "О": "O", "Р": "P", "С": "C", "Т": "T", "Х": "X",
};
const PLATE_RE = /^[ABCEHIKMOPTX]{2}\d{4}[ABCEHIKMOPTX]{2}$/;
const VIN_RE = /^[A-HJ-NPR-Z0-9]{17}$/;

export function normalizePlate(value: string): string | null {
  const compact = value.trim().toUpperCase().replace(/[\s-]+/g, "");
  const candidate = [...compact].map((character) => CYRILLIC_TO_LATIN[character] ?? character).join("");
  return PLATE_RE.test(candidate) ? candidate : null;
}

export function normalizeVin(value: string): string | null {
  const candidate = value.trim().toUpperCase().replace(/[\s-]+/g, "");
  return VIN_RE.test(candidate) ? candidate : null;
}

export function detectQuery(value: string): { kind: "PLATE" | "VIN"; normalized: string } | null {
  const vin = normalizeVin(value);
  if (vin) return { kind: "VIN", normalized: vin };
  const plate = normalizePlate(value);
  if (plate) return { kind: "PLATE", normalized: plate };
  return null;
}

export function languageFor(code: string | undefined, fallback = "uk"): Language {
  if (code?.toLowerCase().startsWith("ru")) return "ru";
  return fallback === "ru" ? "ru" : "uk";
}
